### ISSUE TITLE
"Server-less" gRPC for applications

### ISSUE THREAD

**Issue by ItalyPaleAle** [dapr/dapr#5392]
## Problem we're trying to solve

For Dapr runtimes to communicate with applications, the application needs to have a server running, listening on an open port. Although not an uncommon practice, if we could get rid of this need, the experience of developers using Dapr would be better.

## How can we solve that?

Dapr currently has 2 app channels: HTTP and gRPC. For HTTP, there's not much we can do today, so we'll leave it as-is

For gRPC, due to the duplex nature of the protocol, we (@artursouza and I) have been investigating possible approaches to remove the need for the users' applications to have a running server. The idea is that the Dapr SDKs could always manage connections from the application to Dapr as outbound connections, which would require no listener on the application side.

The application could always create a persistent connection to the runtime at startup. When the runtime needs to send messages to the application, it would re-use that connection.

@artursouza and I completed our initial discovery phase yesterday, and we identified two possible paths to implement this. Although gRPC allows two-way communication, its architecture has a limitation in that only the caller is allowed to initiate new RPCs (either unary or streaming). In this case, we need a solution in which both the caller (app) and callee (server) can send messages over the same connection. We identified 2 possible solutions:

### Using a multiplexer

The most promising one is [hashicorp/yamux](https://github.com/hashicorp/yamux), which is a multiplexer allowing two-way communications over an existing channel (TCP, UDS, or potentially even QUIC in the future).

We have seen at least 2 (open source) implementations that demonstrate how to use gRPC on top of yamux

- https://github.com/chetan/bidi-hello
- https://github.com/dustin-decker/grpc-firewall-bypass

We like this approach for its simplicity. Requiring just a change in the gRPC `Dial/DialWithContext` function to use a connection established outside and that leverages the muxer, it allows then using the existing gRPC invocations we use today, both on the app and on the server. Essentially, this is a change at the gRPC transport layer, which has no visible impacts for applications who continue to work as before. It also allows graceful fallbacks to using a server in the app if needed.

The downside of this approach is that the yamux library is only written in Go. There are community-contributed ports for [Python](https://github.com/xtqxk/pyamux), [JavaScript](https://github.com/th-ch/yamux-js), and [Rust](https://github.com/libp2p/rust-yamux) (_this last one comes from the IPFS project :)_). Because we have SDKs in other programming languages, we would have to implement at least the _client_ part of the [yamux](https://github.com/hashicorp/yamux/blob/master/spec.md) specs in the remaining languages (in theory, using the Rust implementation via language bindings/FFI's could work too, but then we'd have to manage native code which has its own downsides).

Another option for multiplexing could be using QUIC, which would also have the advantage of being quic...ker - but not sure how the support landscape across other programming languages currently is. QUIC is also not TCP-based, so that's a change that goes deeper down the network stack.

### Using a separate gRPC stream for messages from the runtime

In this approach, nothing changes when the app needs to invoke the runtime: it will continue to make outbound connections like it does today.

In addition, at startup the application creates a persistent connection to daprd and invokes a new stream RPC and starts listening for messages.

When daprd needs to send a message to the app (pubsub message, input binding, incoming service invocation/gRPC proxy message, or health probes), it will use this stream and send the request over the stream. The app would also respond in the same stream.

The biggest benefit of this approach is of course that we don't need to change the transport layer, and we continue to use gRPC as-is.

However, there are 2 things to keep in mind:

- We are using a single streaming RPC where the server sends all messages to the app, and the app is expected to respond. This means we will need to implement our own "router" on top of that. _When an app receives a message from the server, where should that be sent: service invocation, input binding, pubsub listener, app health callback?_ Second, we will need to maintain some sort of state correlation, so when the app responds to daprd, daprd knows what that message is in response to. _Is this message an ACK to the pubsub message, a response to the health probe, etc?_ This is going to be more complicated when we start to implement methods that use gRPC streams, as at that point there will be multiple requests and responses that need to be routed (and whose correlation needs to be preserved).
- Second, this is probably going to have a negative performance impact. Over a single HTTP/2 connection, gRPC uses the HTTP/2 multiplexing to have multiple concurrent streams (in this case, "streams" is in the sense that is used in the HTTP/2 specs: it means any active RPC made by gRPC, either unary or stream-based). The default is to allow "at least 100" streams per each HTTP/2 connection. This means that when we use gRPC "normally", there can easily be 100 concurrent requests on the same connection at the same time (e.g. an app could receive up to 100 messages from pubsub in parallel, while it responds to them). When we use a single gRPC stream like in this approach, however, we are limited to having a single message on the wire at the same time. daprd doesn't need to wait for the app to respond before sending another message on the stream, so it doesn't limit scalability to 1 active message at time, but it is likely going to have _some_ negative performance impact. (Of course, if we don't deprecate the current way Dapr works, apps that are sensitive to performance can continue to have their own server, and this would be a tradeoff in the decision of using a server or going "server-less").

**Comment by ItalyPaleAle** [comment]
> This seems to be supported, looking at gRPC docs

Yes, but the case the docs highlight is bidirectional streaming. It is not possible for a gRPC server to create a gRPC client on an incoming connection. Which means that the only way to do 2-way communication is to use a gRPC streaming call (and that's what the second proposed approach is about - with its limitations).

> There's no reason to deprecate the current AppCallback method interface as it enables users to control the communications and network aspects (with security being top of mind here) and provides users with a choice that best fits their needs.

Agree. I was not suggesting we deprecate the current AppCallback by the way.

**Comment by ItalyPaleAle** [comment]
After investigating further, I made a lot of progress and now have a working POC for this :) 

https://github.com/ItalyPaleAle/dapr-firewall-poc

I ended up implementing in a completely different way from the 2 described above, and I think this is possibly the simplest one too. 

At a high level, the app now connects to daprd via gRPC (as always) and asks daprd to create an ephemeral TCP listener on a random port. The app then establishes a TCP connection to that port, and then the connection is turned into a gRPC server in the app, and a gRPC client connection in daprd.

Just like all other solutions, it does require work in each SDK, but it doesn't depend on any external library (like yamux) and doesn't require implementing complicated routing logic (like if we were to use a bidirectional gRPC stream as transport layer).

It should also have pretty good performance: by default, gRPC allows up to 100 concurrent "streams" per connection (ie. per TCP connection), and should not have (additional, besides what's baked into TCP itself) problems with head-of-line blocking (like using gRPC streams as transports could have).

The repo linked above has the full code of the POC, including steps to run it locally. It works E2E, but there's still some work left to do especially with error handling and recovering from failures (such as dropped connections).

CC: @yaron2 @artursouza

**Comment by philliphoff** [comment]
@ItalyPaleAle asked me about the .NET SDK side of this issue and I spent a little time investigating its feasibility.  While I don't know yet for certain, it appears that the underlying .NET HTTP stack doesn't have Go's flexibility in allowing a gRPC client connection to establish gRPC server host. (See:  [grpc-dotnet#486](https://github.com/grpc/grpc-dotnet/issues/486))

Given that, another approach (at least for .NET and, perhaps, JS as well), is to have a Go-based intermediary that ships with the SDKs, does the gRPC "dance", and proxies requests back to the SDK which itself is hosting a "normal" gRPC server.  Since that intermediary is on the same machine (or, at least, the same network), it should be able to make a connection back to the SDK's server.

```mermaid
sequenceDiagram
Application->>Application: Start "public" HTTP host
Application->>SDK: Start Server
SDK->>SDK: Start "private" HTTP host and gRPC server
SDK->>Go Proxy: Start server
Go Proxy->>Dapr: Start server
Dapr->>Dapr: Start HTTP host and gRPC server
Go Proxy->>Dapr: Establish TCP connection to gRPC server
Go Proxy->>Go Proxy: Add gRPC server to client connection
Internet->>Dapr: Deliver pub-sub message via binding
Dapr->>Go Proxy: Deliver pub-sub message via gRPC
Go Proxy->>SDK: Deliver pub-sub message via gRPC
SDK->>Application: Deliver pub-sub message via callback
```

Admittedly, introducing a new component (and one that's platform specific) significantly increases the complexity and test matrix of the SDKs.  However, the SDKs can easily shield applications from such implementation details (by managing the "private" gRPC server) such that, if and when the HTTP stacks better support this scenario, they can be swapped without further application changes.
